h2/src/main/org/h2/command/dml/MergeUsing.java

```diff
@@ -0,0 +1,496 @@
+/*
+ * Copyright 2004-2017 H2 Group. Multiple-Licensed under the MPL 2.0,
+ * and the EPL 1.0 (http://h2database.com/html/license.html).
+ * Initial Developer: H2 Group
+ */
+package org.h2.command.dml;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import org.h2.api.ErrorCode;
+import org.h2.api.Trigger;
+import org.h2.command.CommandInterface;
+import org.h2.command.Prepared;
+import org.h2.engine.Right;
+import org.h2.engine.Session;
+import org.h2.expression.ConditionAndOr;
+import org.h2.expression.Expression;
+import org.h2.expression.ExpressionVisitor;
+import org.h2.message.DbException;
+import org.h2.result.ResultInterface;
+import org.h2.result.Row;
+import org.h2.result.RowImpl;
+import org.h2.table.Column;
+import org.h2.table.Table;
+import org.h2.table.TableFilter;
+import org.h2.util.New;
+import org.h2.util.StatementBuilder;
+import org.h2.value.Value;
+
+/**
```

Comment:
@grandinj Please read this lengthy module comment - it explains a lot about this class and why it implemented like this... also one potential issue mentioned at the end (aka a future bug report)

Reply:
Added fix for potential bug report